### PR TITLE
Be less aggressive with `DroplessArena`/`TypedArena` growth.

### PR DESCRIPTION
**Pull request by nnethercote**
`DroplessArena` and `TypedArena` use an aggressive growth strategy: each chunk is double the size of the previous. The first chunk is 4 KiB, the second is 8 KiB, the third is 16 KiB, and so on, indefinitely. (Note that each new chunk is *added* to the arena; old chunks are kept, and there is no reallocation of chunks occurring.) DHAT profiles show that sometimes this results in large chunks (e.g. 16-128 MiB) that are barely filled.
    
This commit changes things so that the doubling stops at 2 MiB. This is large enough that chunk allocations are still rare (you might get 100s instead of 10s of them) but avoids lots of unused space in the worst case. It makes the same change to `TypedArena`, too.

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
I wonder if we should increase the first chunk, too - 4 KB seems quite small to me, most crates use much more I imagine? It would be nice to get statistics I guess on how much memory each arena uses in total. In any case, it probably doesn't matter that much, I'm guessing the arena allocations are all large enough that the allocator falls back to mmap or whatever anyway.

One thing that I had wanted to experiment with is using (at least on Linux) huge pages for our arenas with the intent of reducing TLB misses - but I haven't had time to investigate.

Anyway, this is all just throwing thoughts out there, not things that need to happen in this PR.

File: src/libarena/lib.rs [hunk 1]
```diff
@@ -99,6 +99,7 @@ impl<T> TypedArenaChunk<T> {
 }
 
 const PAGE: usize = 4096;
+const BIG_CHUNK: usize = 1024 * 1024;
```
Comment (the8472) [inline src/libarena/lib.rs:102]: Have you tried 2MB? That would match the huge page size on x86 and benefit from THP optimization, at least on `transparent_hugepage/enabled = always` systems.

**Comment by nnethercote** [comment]
> I wonder if we should increase the first chunk, too - 4 KB seems quite small to me

I think 4 KiB is a good choice. We don't want smaller than that because it's commonly the page size. As for larger, in my profiles I've seen that it's pretty common to have `TypedArena`s that are barely used, e.g. only 64 bytes out of 4 KiB are used. I'm going to see if I can avoid some of those today, but in general the design is nicely flexible -- if the arena is barely used you only waste ~4 KiB but if it's used a lot it quickly scales up to 1 or 2 MiB, and then with the changes in this PR we don't get overly large chunks that can waste lots of space.

File: src/libarena/lib.rs (hunk 1 above)
Comment (nnethercote) [reply src/libarena/lib.rs:102]: I haven't, but I'm happy to change it to 2 MiB. I'll add a comment explaining why.

**Comment by nnethercote** [comment]
The max RSS results are noisy as usual, but I didn't expect an improvement there. The wasted space from the overly-large chunks shouldn't contribute to RSS, because the memory is never touched.

So why bother with this, then? Two reasons:
- The cycle counts in the perf results look like a slight improvement overall, and maybe likewise with wall-times though it's harder to tell.
- I'm starting a push on memory usage and these overly-large chunks clog up DHAT's output. Getting rid of them will make it easier for me to make subsequent progress.

**Comment by nnethercote** [comment]
I have updated to use 2 MiB as the limit.

**Comment by the8472** [comment]
I'm not sure if this is an issue, but changing from exponential to linear means that the number of memory mappings increase (assuming the underlying malloc does mmap for large allocations) and there's a limit on those (`vm.max_map_count = 65530` on my system). That means it lowers the maximum that can be allocated via these arenas to at most 130GB, likely less.
That ought to be enough for anyone, but still might be a noteworthy change.

**Comment by nnethercote** [comment]
Eight days with no review response. Let's try a different reviewer.

r? @oli-obk

File: src/libarena/lib.rs [hunk 2]
```diff
@@ -221,16 +227,18 @@ impl<T> TypedArena<T> {
...
-                    loop {
+                    if new_capacity < HUGE_PAGE / elem_size {
                         new_capacity = new_capacity.checked_mul(2).unwrap();
-                        if new_capacity >= currently_used_cap + n {
-                            break;
-                        }
                     }
+
+                    // Also ensure that this chunk can fit `n`.
+                    new_capacity = cmp::max(n, new_capacity);
```
Comment (oli-obk) [inline src/libarena/lib.rs:239]: I think this should happen just before the `TypedArenaChunk::new` and the `cmp::max` below should be removed. This way we guarantee that it happens for both paths and documents this happening the other branch below, too.

Same thing for the other arena

File: src/libarena/lib.rs [hunk 3]
```diff
@@ -211,6 +216,7 @@ impl<T> TypedArena<T> {
     #[cold]
     fn grow(&self, n: usize) {
         unsafe {
+            let elem_size = cmp::max(1, mem::size_of::<T>());
```
Comment (oli-obk) [inline src/libarena/lib.rs:219]: leave a comment here that we need the element size in order to go from byte sizes of `HUGE_PAGE` and `PAGE` to element counts obtained from `capacity` and used for creating new typed chunks.

**Comment by oli-obk** [comment]
just some nits, r=me with those resolved

File: src/libarena/lib.rs (hunk 2 above)
Comment (nnethercote) [reply src/libarena/lib.rs:239]: Good idea.

**Comment by RalfJung** [comment]
(I am going to assume that this is deliberately not rollup=never, and include this in rollups. If that is undesired, please mark PR as rollup=never.)

**Comment by nnethercote** [comment]
Correct, the perf effect is small enough that this would be fine in a rollup.

**Comment by RalfJung** [comment]
Something failed both #72226 and #72228, and there are only two PRs left in the intersection: this one, and https://github.com/rust-lang/rust/pull/72090. The latter is an unlikely candidate.

So could you check those errors and see if they might be caused by this PR?

**Comment by CAD97** [comment]
My (complete) guess would be that some code is still expecting the capacity to continue doubling after hitting the `HUGE_PAGE` size.

Actually, if I didn't miss anything when glancing at the diff, I think your code just outright stops increasing the size of the arena once it fills a huge page? If that's correct, that's obviously problematic when an arena attempts to increase past that size.

**Comment by nnethercote** [comment]
> So could you check those errors and see if they might be caused by this PR?

It's not *obviously* the fault of this PR. I'm having trouble imagining how this PR would cause lots of test failures only on Windows builds. This is platform-independent allocator code... if the code was bad, I would expect failures on all platforms.

Can we wait for it to land by itself? That would give a clear signal whether this PR is the problem. (It's a shame we can't schedule a try run on specific platforms in advance of merging.)

**Comment by RalfJung** [comment]
It could also be 32bit-only, and the Windows builders are just the first to fail.

**Comment by RalfJung** [comment]
Anyway let's make sure to not roll this up until we got confirmation.
@bors rollup=never

**Comment by RalfJung** [comment]
Strange, looks like all the PRs in the intersection of those two rollups have landed now...

**Comment by bill-myers** [comment]
Doesn't this potentially cause quadratic instead of linear behavior?

I think you need to enforce multiplicative growth in all cases, although a multiplicative factor lower than 2 can be used.

**Pull request merged**